motionplan/plan.go

```diff
@@ -0,0 +1,226 @@
+// A Point with X as the longitude and Y as the latitude
+// An orientation using the heading as the theta in an OrientationVector with Z=1.
+func NewGeoPlan(p Plan, geoOrigin *spatialmath.GeoPose) (Plan, error) {
+	plan, ok := p.(*rrtPlan)
+	if !ok {
+		return nil, errBadPlanImpl
+	}
+	newPath := make([]PathStep, 0, len(plan.Path()))
+	for _, step := range plan.Path() {
+		newStep := make(PathStep)
+		for frame, pif := range step {
+			pose := pif.Pose()
+			geoPose := spatialmath.PoseToGeoPose(geoOrigin, pose)
+			heading := math.Mod(math.Abs(geoPose.Heading()-360), 360)
+			o := &spatialmath.OrientationVectorDegrees{OZ: 1, Theta: heading}
+			smuggledGeoPose := spatialmath.NewPose(r3.Vector{X: geoPose.Location().Lng(), Y: geoPose.Location().Lat()}, o)
+			newStep[frame] = referenceframe.NewPoseInFrame(pif.Parent(), smuggledGeoPose)
+		}
+		newPath = append(newPath, newStep)
+	}
+	return &rrtPlan{
+		traj:  plan.traj,
+		path:  newPath,
+		nodes: plan.nodes,
+	}, nil
+}
+
+// Trajectory is a slice of maps describing a series of Inputs for a robot to travel to in the course of following a Plan.
+// Each item in this slice maps a Frame's name (found by calling frame.Name()) to the Inputs that Frame should be modified by.
+type Trajectory []map[string][]referenceframe.Input
+
+// GetFrameInputs is a helper function which will extract the waypoints of a single frame from the map output of a trajectory.
+func (traj Trajectory) GetFrameInputs(frameName string) ([][]referenceframe.Input, error) {
+	solution := make([][]referenceframe.Input, 0, len(traj))
+	for _, step := range traj {
+		frameStep, ok := step[frameName]
+		if !ok {
+			return nil, fmt.Errorf("frame named %s not found in trajectory", frameName)
+		}
+		solution = append(solution, frameStep)
+	}
+	return solution, nil
+}
+
+// String returns a human-readable version of the trajectory, suitable for debugging.
+func (traj Trajectory) String() string {
+	var str string
+	for _, step := range traj {
+		str += "\n"
+		for frame, input := range step {
+			if len(input) > 0 {
+				str += fmt.Sprintf("%s: %v\t", frame, input)
+			}
+		}
+	}
+	return str
+}
+
+// EvaluateCost calculates a cost to a trajectory as measured by the given distFunc Metric.
+func (traj Trajectory) EvaluateCost(distFunc ik.SegmentMetric) float64 {
+	var totalCost float64
+	last := map[string][]referenceframe.Input{}
+	for _, step := range traj {
+		for frame, inputs := range step {
+			if len(inputs) > 0 {
+				if lastInputs, ok := last[frame]; ok {
+					cost := distFunc(&ik.Segment{StartConfiguration: lastInputs, EndConfiguration: inputs})
+					totalCost += cost
+				}
+				last[frame] = inputs
+			}
+		}
+	}
+	return totalCost
+}
+
+// Path is a slice of PathSteps describing a series of Poses for a robot to travel to in the course of following a Plan.
+type Path []PathStep
+
+func newRelativePath(solution []node, sf *solverFrame) (Path, error) {
```

Comment:
Should this be a private method on solverFrame?

Reply:
Also, there doesn't seem to be anything `relative`-specific about this function

Reply:
Youre right this was confusing.  I've cleaned it up. I dont think it should be a method on solver frame I would rather it be its own thing